server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -108,14 +147,62 @@ public CoordinatorStats run(DruidCoordinator coordinator, DruidCoordinatorRuntim
     // Remove over-replication
     stats.accumulate(drop(loadStatus, segment, params));
 
-
     return stats;
   }
 
-  private int assign(
+  @Nullable
+  private static ServerHolder getPrimaryHolder(
+      final DruidCluster cluster,
+      final Map<String, Integer> tieredReplicants,
+      final BalancerStrategy strategy,
+      final DataSegment segment,
+      final Predicate<ServerHolder> serverHolderPredicate
+  )
+  {
+    ServerHolder topCandidate = null;
+
+    for (final Map.Entry<String, Integer> entry : tieredReplicants.entrySet()) {
+      final int expectedReplicantsInTier = entry.getValue();
+      if (expectedReplicantsInTier <= 0) {
+        continue;
+      }
+
+      final String tier = entry.getKey();
+
+      final MinMaxPriorityQueue<ServerHolder> serverQueue = cluster.getHistoricalsByTier(tier);
+      if (serverQueue == null) {
+        log.makeAlert("Tier[%s] has no servers! Check your cluster configuration!", tier).emit();
+        continue;
+      }
+
+      final ServerHolder candidate = strategy.findNewSegmentHomeReplicator(
```

Comment:
If I'm reading this correctly, this is doing a LOT more here compared to what it was previously

Reply:
I think it should not be that bad. If the `findNewSegmentHomeReplicator` method call is of concern, then this change will make _N - 1_ more calls where _N_ is the number of tiers. If we have about 2 tiers per rule, this will make one more call.